examples/vision/supervised-contrastive-learning.py

```diff
@@ -0,0 +1,206 @@
+  )
+  return model
+
+"""## Experiment 1: Train the baseline classification model 
+
+In this experiment, a baseline classifier is trained normally, i.e., the
+encoder and the classifier parts are trained together as a single model
+to minimize cross-entropy loss.
+"""
+
+encoder = create_encoder()
+classifier = create_classifier(encoder)
+classifier.summary()
+
+history = classifier.fit(
+  data_generator.flow(x_train, y_train, batch_size),
+  epochs=num_epochs)
+
+accuracy = classifier.evaluate(data_generator.flow(x_test, y_test))[1]
+print(f'Test accuracy: {round(accuracy * 100, 2)}%')
+
+"""We get to ~76.5% test accuracy.
+
+## Experiment 2: Use supervised contrastive learning
+In this experiment, the model is trained in two phases. In the first phase,
+the encoder is pretrained to optimize the supervised contrastive loss,
+described in [Prannay Khosla et al.](https://arxiv.org/abs/2004.11362).
+In the second phase, the classifier is trained using the trained encoder with
+its weights freezed; only the weights of fully-connected layers with the
+softmax are optimized.
+
+### 1. Supervised contrastive learning loss function
+"""
+
+class SupervisedContrastiveLoss(keras.losses.Loss):
+  def __init__(self, temperature=1, name=None):
+    super(SupervisedContrastiveLoss,self).__init__(name=name)
+    self.temperature = temperature
+    
+  def __call__(self, labels, feature_vectors, sample_weight=None):
+    # Normalize feature vectors
+    feature_vectors_normalized = tf.math.l2_normalize(feature_vectors, axis=1)
+    # Compute logits 
+    logits = tf.divide(
+        tf.matmul(
+            feature_vectors_normalized, 
+            tf.transpose(feature_vectors_normalized)),
+        temperature
+    )
+    return tfa.losses.npairs_loss(tf.squeeze(labels), logits)
+
+def add_projection_head(encoder):
+  inputs = keras.Input(shape=input_shape)
+  features = encoder(inputs)
+  outputs = layers.Dense(projection_units, 'relu')(features)
+  outputs = layers.BatchNormalization()(features)
+  model = keras.Model(inputs=inputs, outputs=outputs, 
+                      name='cifar-encoder_with_projection-head')
+  return model
+
+"""### 2. Pretrain the encoder"""
+
+encoder = create_encoder()
+
+encoder_with_projection_head = add_projection_head(encoder)
+encoder_with_projection_head.compile(
+    optimizer=keras.optimizers.Adam(learning_rate),
+    loss=SupervisedContrastiveLoss(temperature)
+)
+
+encoder_with_projection_head.summary()
+
+history = encoder_with_projection_head.fit(
+  data_generator.flow(x_train, y_train, batch_size),
+  epochs=num_epochs)
+
+"""### 3. Train the classifier with the frozen encoder"""
+
+classifier = create_classifier(encoder, trainable=False)
+
+history = classifier.fit(
+  data_generator.flow(x_train, y_train, batch_size), 
+  epochs=num_epochs,
+  )
+
+accuracy = classifier.evaluate(data_generator.flow(x_test, y_test))[1]
+print(f'Test accuracy: {round(accuracy * 100, 2)}%')
+
+"""We get to ~81.3% test accuracy."""
```

Comment:
Please add a comment about comparing the two methods, and maybe some pointers to which factors affect when to use SCL and when not to use it.

Reply:
@fchollet - I added a 'conclusion' section at the end. LMK if this what you had in mind.